src/main/java/com/google/sps/servlets/EventFreeTimeServlet.java

```diff
@@ -0,0 +1,198 @@
+    } catch (InterruptedException e2) {
+      e2.printStackTrace();
+      response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+      return;
+    }
+
+    usersCredentials.forEach(userCredential -> {
+      try {
+        createAndAddBatchRequest(batchRequest, callback, userCredential, startDateTimeWithShift, endDateTimeWithShift);
+        batchRequest.execute();
+        returnFreeTimesFromInterval(startDateTimeWithShift.getValue(), endDateTimeWithShift.getValue(), duration, busyTimesForAttendees, response);
+        return;
+      } catch (IOException e1) {
+        // TODO Auto-generated catch block
+        e1.printStackTrace();
+      } catch (GeneralSecurityException e1) {
+        // TODO Auto-generated catch block
+        e1.printStackTrace();
+      }
+      return;
+    });
+
+  }
+
+  private void createAndAddBatchRequest(BatchRequest batchRequest, JsonBatchCallback<FreeBusyResponse> callback,
+      Credential userCredential, DateTime start, DateTime end) throws IOException, GeneralSecurityException {
+    Calendar userCalendar = loadCalendarClient(userCredential);
+    FreeBusyRequest fBusyRequest = new FreeBusyRequest().setTimeMin(start).setTimeMax(end)
+        .setItems(Collections.singletonList(new FreeBusyRequestItem().setId("primary")));
+    HttpRequest request = userCalendar.freebusy().query(fBusyRequest).buildHttpRequest();
+    batchRequest.queue(request, FreeBusyResponse.class, GoogleJsonErrorContainer.class, callback);
+  }
+
+  private void returnFreeTimesFromInterval(Long start, Long end, Long duration, Set<Time> busyTimes, HttpServletResponse response)
+      throws IOException {
+    List<Time> freeTimes = new ArrayList<>();
+    Long shift = Long.valueOf(1000 * 60 * 15);
+    Long currentStart = start;
+    Long currentEnd = start + duration * 60 * 1000;
+    while (currentEnd <= end) {
+      boolean overlap = false;
+      for (Time busyTime : busyTimes) {
+        if (busyTime.getEnd() <= currentStart)
+          continue;
+        if (busyTime.getStart() >= currentEnd)
+          break; // Ordered by start, so we can break early
+        overlap = true;
+        break;
+      }
+      if (!overlap)
+        freeTimes.add(new Time(currentStart, currentEnd));
+      currentStart = currentStart + shift;
+      currentEnd += shift;
+    }
+    response.setContentType("application/json");
+    response.getWriter().println(new Gson().toJson(freeTimes));
+
+    return ;
+  }
+
+  private List<Credential> getCredentialsFromUserList(List<String> userIds) throws InterruptedException {
+    List<Credential> usersCredential = new ArrayList<>();
+    ExecutorService executorService = Executors.newCachedThreadPool(ThreadManager.currentRequestThreadFactory());
+
+    userIds.forEach(userId -> {
+      executorService.execute(ThreadManager.createThreadForCurrentRequest(() -> {
+        try {
+          usersCredential.add(flow.loadCredential(userId));
```

Comment:
This looks also dangerous because ArrayList is not thread safe. Maybe better use the invokeAll functionality and deal with a list of futures which you can easily handle one after another without any concurrency issues. See, e.g., here: https://www.baeldung.com/java-executor-service-tutorial